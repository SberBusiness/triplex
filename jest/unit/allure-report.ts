import {JasmineAllureReporter} from 'allure-jasmine';
import {Status, TestResult} from 'allure-js-commons';

const reporter = new JasmineAllureReporter({
    resultsDir: 'tests/out/unit/allure/allure-results',
    testMapper: (result: TestResult) => {
        if (result.status == Status.SKIPPED) result.fullName = `(WAS SKIPPED) ${result.fullName}`;
        return result;
    },
});

// @ts-ignore
jasmine.getEnv().addReporter(reporter);

export const allure = reporter.getInterface();
