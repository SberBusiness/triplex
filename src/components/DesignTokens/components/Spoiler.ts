import {TDesignTokenValue, TDesignTokenValues} from '@sberbusiness/triplex/components/DesignTokens/types/DesignTokenTypes';

// Название токенов компонента Spoiler.
export const designTokensComponentsSpoilerKeys = ['Caret_Color'] as const;
// Тип, содержащий названия токенов компонента Spoiler.
export type TDesignTokensComponentsSpoilerKeys = (typeof designTokensComponentsSpoilerKeys)[number];
// Тип, содержащий названия токенов компонента Spoiler и их значения.
export type TDesignTokensComponentsSpoilerValue = Record<TDesignTokensComponentsSpoilerKeys, TDesignTokenValue>;
// Тип, содержащий названия токенов компонента Spoiler и их значения в светлой и темной теме.
export type TDesignTokensComponentsSpoilerValues = Record<TDesignTokensComponentsSpoilerKeys, TDesignTokenValues>;
// Тип локальных токенов компонента Spoiler.
export type TDesignTokensComponentsSpoiler = {Spoiler: TDesignTokensComponentsSpoilerValue};

// Токены компонента Spoiler в светлой и темной темах.
export const Spoiler_Tokens: TDesignTokensComponentsSpoilerValues = {
    Caret_Color: [{ref: 'Basic.700'}, {ref: 'Neutral.700'}], // var(--triplex-Spoiler-Caret_Color)
};
