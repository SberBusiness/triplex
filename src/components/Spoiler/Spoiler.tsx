import React from 'react';
import {CaretdownSrvxIcon16} from '@sberbusiness/icons/CaretdownSrvxIcon16';
import {KeyDownListener} from '@sberbusiness/triplex/components/KeyDownListener/KeyDownListener';
import {EVENT_KEY_CODES} from '@sberbusiness/triplex/utils/keyboard';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';
import {Button} from '@sberbusiness/triplex/components/Button/Button';
import {EButtonTheme, EButtonSize} from '@sberbusiness/triplex/components/Button/enums';
import {uniqueId} from '@sberbusiness/triplex/utils/uniqueId';

/** Базовые свойства компонента Spoiler. */
interface ISpoilerBaseProps extends React.HTMLAttributes<HTMLDivElement> {
    /** Текст раскрытия содержимого. */
    labelExpand: string;
    /** Текст скрытия содержимого. */
    labelCollapse?: string;
    /** Обработчик скрытия/раскрытия. */
    onToggle?: (expanded: boolean) => void;
    /** Элемент правого блока. */
    rightBlock?: JSX.Element;
}

/** Свойства контролируемого Spoiler. */
interface ISpoilerControlledProps extends ISpoilerBaseProps {
    /** Контролируемое состояние скрыт/раскрыт. */
    expanded: boolean;
    /** Контролирующая функция скрытия/раскрытия. */
    toggle: (nextExpanded: boolean) => void;
}

/** Свойства неконтролируемого Spoiler. */
interface ISpoilerUncontrolledProps extends ISpoilerBaseProps {
    /** Контролируемое состояние скрыт/раскрыт. */
    expanded?: never;
    /** Контролирующая функция скрытия/раскрытия. */
    toggle?: never;
}

/** Комбинированные свойства компонента Spoiler. */
type TSpoilerProps = ISpoilerControlledProps | ISpoilerUncontrolledProps;

/** Состояния компонента Spoiler. */
interface ISpoilerState {
    /** Раскрыт или нет. */
    isExpanded: boolean;
    /** Контролируемый или нет. */
    isControlled: boolean;
}

/** Компонент "Спойлер", используется для раскрытия внутреннего содержимого. */
export class Spoiler extends React.Component<TSpoilerProps, ISpoilerState> {
    public static displayName = 'Spoiler';

    node: HTMLButtonElement | null = null;

    // Уникальный id, для передачи a11y aria-атрибутов.
    private instanceId = `Spoiler-${uniqueId()}`;

    public state = {
        isControlled: this.props.expanded !== undefined,
        isExpanded: !!this.props.expanded,
    };

    public render(): JSX.Element {
        const {children, expanded, onToggle, toggle, className, labelExpand, labelCollapse, rightBlock, ...divHTMLAttributes} = this.props;
        const {isExpanded, isControlled} = this.state;
        const open = isControlled ? !!expanded : isExpanded;

        const classContent = classnames('cssClass[content]', {'cssClass[hidden]': !open});
        const classDownIcon = classnames('cssClass[icon]', 'cssClass[animationRotate180]', {'cssClass[active]': open});

        const toggleLabel = (open && labelCollapse) || labelExpand;

        return (
            <KeyDownListener eventKeyCode={EVENT_KEY_CODES.ENTER} onMatch={this.toggleContentByKeyDown}>
                <div {...divHTMLAttributes} className={classnames('cssClass[spoiler]', className)}>
                    <div className={'cssClass[head]'}>
                        <Button
                            aria-controls={this.instanceId}
                            aria-expanded={open}
                            className="cssClass[toggleButton]"
                            theme={EButtonTheme.LINK}
                            size={EButtonSize.SM}
                            onClick={this.toggleContent}
                            ref={this.setRef}
                        >
                            {toggleLabel}
                            <CaretdownSrvxIcon16 className={classDownIcon} />
                        </Button>
                        {rightBlock}
                    </div>
                    <div className={classContent} id={this.instanceId}>
                        {children}
                    </div>
                </div>
            </KeyDownListener>
        );
    }

    public componentDidUpdate(prevProps: TSpoilerProps): void {
        const {expanded, onToggle} = this.props;
        const {expanded: prevExpanded} = prevProps;

        if (expanded !== prevExpanded) {
            onToggle?.(!!expanded);
        }
    }

    /** Обработчик открытия и закрытия спойлера. */
    private toggleContent = (): void => {
        const {onToggle, toggle, expanded} = this.props;
        const {isControlled, isExpanded} = this.state;

        if (isControlled) {
            toggle?.(!expanded);
        } else {
            this.setState(
                (prevState) => ({isExpanded: !prevState.isExpanded}),
                () => onToggle?.(!isExpanded)
            );
        }
    };

    /** Обработчик открытия и закрытия спойлера по нажатию клавиши. */
    private toggleContentByKeyDown = (event: KeyboardEvent): void => {
        if (document.activeElement === this.node) {
            event.preventDefault();
            this.toggleContent();
        }
    };

    private setRef = (ref: HTMLButtonElement): void => {
        this.node = ref;
    };
}
