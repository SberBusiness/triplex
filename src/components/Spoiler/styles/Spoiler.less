@import 'src/styles/style';

.spoiler {
    display: flex;
    flex-direction: column;

    .head {
        display: flex;
        justify-content: space-between;

        .icon {
            margin-left: @metric-4;

            path {
                fill: var(--triplex-Spoiler-Caret_Color);
            }

            &.animationRotate180 {
                transition: transform 0.3s ease-in-out;

                &.active {
                    transform: rotate(-180deg);
                }
            }
        }

        .toggleButton {
            margin-left: -2px;
        }
    }

    .content {
        margin-top: @ufs-theme-base*2;

        &.hidden {
            display: none;
        }
    }
}
