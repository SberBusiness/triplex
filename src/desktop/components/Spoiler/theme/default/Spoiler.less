@import '../../../../common/theme/default';

.spoiler {
    display: flex;
    flex-direction: column;

    .head {
        display: flex;
        justify-content: space-between;

        .icon {
            margin-left: @metric-4;

            &.animationRotate180 {
                transition: transform 0.25s ease-in-out;

                &.active {
                    transform: rotate(-180deg);
                }
            }
        }
    }

    .content {
        margin-top: @ufs-theme-base*2;

        &.hidden {
            display: none;
        }
    }
}
