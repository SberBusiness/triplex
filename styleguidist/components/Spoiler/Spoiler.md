```jsx
<Spoiler labelExpand="Развернуть" labelCollapse="Свернуть">
    Content
</Spoiler>
```

### Spoiler Controlled

```jsx
import {ComponentControlPanel} from '../../common/components/ComponentControlPanel/ComponentControlPanel';

const [expanded, setExpanded] = React.useState(false);

const renderControlPanel = () => (
    <ComponentControlPanel>
        <ComponentControlPanel.Checkbox checked={expanded} setChecked={setExpanded}>
            Expanded
        </ComponentControlPanel.Checkbox>
    </ComponentControlPanel>
);

<>
    {renderControlPanel()}
    <Spoiler labelExpand="Развернуть" labelCollapse="Свернуть" expanded={expanded} toggle={setExpanded}>
        Content
    </Spoiler>
</>
```
