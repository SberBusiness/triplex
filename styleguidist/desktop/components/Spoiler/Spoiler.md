### Spoiler Basic

```jsx
import React from 'react';
import {Col} from '@sberbusiness/triplex/desktop/components/Col/Col';
import {Field} from '@sberbusiness/triplex/desktop/components/Field/Field';
import {SubRow} from '@sberbusiness/triplex/desktop/components/SubRow/SubRow';
import {SubLabel} from '@sberbusiness/triplex/desktop/components/SubLabel/SubLabel';
import {SubValue} from '@sberbusiness/triplex/desktop/components/SubValue/SubValue';

const renderSubRow = (label, index) => {
    return (
        <SubRow key={index}>
            <Col>
                <Field alignLabel>
                    <Col size={2}>
                        <SubLabel>{label}</SubLabel>
                    </Col>
                    <Col size={2}>
                        <SubValue>{label}</SubValue>
                    </Col>
                </Field>
            </Col>
        </SubRow>
    );
};

const renderSpoilerContent = (count) => Array(count).fill().map((value, index) => renderSubRow('Текст в 1 строку', index));

<>
    <Spoiler labelExpand="Развернуть" labelCollapse="Свернуть">
        {renderSpoilerContent(4)}
    </Spoiler>
</>
```

### Spoiler Controlled

```jsx
import React, {useState} from 'react';
import {Col} from '@sberbusiness/triplex/desktop/components/Col/Col';
import {Field} from '@sberbusiness/triplex/desktop/components/Field/Field';
import {SubRow} from '@sberbusiness/triplex/desktop/components/SubRow/SubRow';
import {SubLabel} from '@sberbusiness/triplex/desktop/components/SubLabel/SubLabel';
import {SubValue} from '@sberbusiness/triplex/desktop/components/SubValue/SubValue';

const [expanded, setExpanded] = useState(false);

const renderSubRow = (label, index) => {
    return (
        <SubRow key={index}>
            <Col>
                <Field alignLabel>
                    <Col size={2}>
                        <SubLabel>{label}</SubLabel>
                    </Col>
                    <Col size={2}>
                        <SubValue>{label}</SubValue>
                    </Col>
                </Field>
            </Col>
        </SubRow>
    );
};

const renderSpoilerContent = (count) => Array(count).fill().map((value, index) => renderSubRow('Текст в 1 строку', index));

const handleChange = (event) => setExpanded(event.target.checked);

const renderControls = () => {
    return (
        <div style={{marginBottom: '16px'}}>
            <label style={{display: 'flex'}}>
                <input type="checkbox" checked={expanded} onChange={handleChange} />
                Expanded
            </label>
        </div>
    );
};

<>
    {renderControls()}
    <Spoiler labelExpand="Развернуть" labelCollapse="Свернуть" expanded={expanded} toggle={setExpanded}>
        {renderSpoilerContent(4)}
    </Spoiler>
</>
```
