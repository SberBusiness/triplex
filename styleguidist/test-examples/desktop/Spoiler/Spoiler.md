```jsx
import React, {useState} from 'react';
import ComponentOptions from '../../../common/components/ComponentOptions/ComponentOptions';
import ComponentPreview from '../../../common/components/ComponentPreview/ComponentPreview';
import {Spoiler} from '@sberbusiness/triplex/desktop/components/Spoiler/Spoiler';

const [expanded, setExpanded] = useState(false);

const checkboxOptions = [
    {
        id: 'expanded',
        label: 'Expanded',
        checked: expanded,
        onChange: setExpanded,
    },
];

<>
    <ComponentOptions checkboxOptions={checkboxOptions} />
    <ComponentPreview>
        <Spoiler labelExpand="Развернуть" labelCollapse="Свернуть" expanded={expanded} toggle={setExpanded}>
            Шаблонный текст
        </Spoiler>
    </ComponentPreview>
</>
```