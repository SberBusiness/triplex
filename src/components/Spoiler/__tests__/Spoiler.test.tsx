import React from 'react';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import {Spoiler} from '../Spoiler';
import {allure} from '@jest/unit/allure-report';

jest.mock('@sberbusiness/icons/SpinnersmallAniIcon20', () => ({
    SpinnersmallAniIcon20: 'svg',
}));

jest.mock('@sberbusiness/icons/SpinnersmallwhiteAniIcon20', () => ({
    SpinnersmallwhiteAniIcon20: 'svg',
}));

jest.mock('@sberbusiness/icons/CaretdownSrvxIcon16', () => ({
    CaretdownSrvxIcon16: 'svg',
}));

describe('Spoiler', () => {
    beforeEach(() => {
        allure.feature('Spoiler');
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <Spoiler labelExpand="Развернуть">
                    <span>123456</span>
                </Spoiler>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('calls onToggle', () => {
        const handleToggle = jest.fn();
        const spoiler = shallow(<Spoiler labelExpand="Развернуть" onToggle={handleToggle} />);
        spoiler
            .childAt(0)
            .childAt(0)
            .childAt(0)
            .simulate('click');
        expect(handleToggle).toBeCalledTimes(1);
    });

    it('changes visibility state', () => {
        const spoiler = shallow(<Spoiler labelExpand="Развернуть" />);
        expect(spoiler.state('isExpanded')).toBeFalsy();
        spoiler
            .childAt(0)
            .childAt(0)
            .childAt(0)
            .simulate('click');
        expect(spoiler.state('isExpanded')).toBeTruthy();
        spoiler
            .childAt(0)
            .childAt(0)
            .childAt(0)
            .simulate('click');
        expect(spoiler.state('isExpanded')).toBeFalsy();
    });

    it('takes the initial state of visibility', () => {
        const spoiler = shallow(<Spoiler labelExpand="Развернуть" expanded={true} toggle={jest.fn()} />);

        expect(spoiler.state('isExpanded')).toBeTruthy();
    });
});
