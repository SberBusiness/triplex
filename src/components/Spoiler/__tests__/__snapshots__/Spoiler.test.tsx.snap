// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spoiler renders correctly 1`] = `
<div
  className="cssClass[spoiler]"
>
  <div
    className="cssClass[head]"
  >
    <button
      className="cssClass[button] cssClass[link] cssClass[SM] cssClass[toggleButton]"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      type="button"
    >
      <span
        className="cssClass[content]"
      >
        Развернуть
        <svg
          className="cssClass[icon] cssClass[animationRotate180]"
        />
      </span>
    </button>
  </div>
  <div
    className="cssClass[content] cssClass[hidden]"
  >
    <span>
      123456
    </span>
  </div>
</div>
`;
